[ERM UI] Add downstream domain UI work

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[JIRA Ticket](https://dimagi-dev.atlassian.net/browse/SS-103)

**Review commit by commit**

This PR is responsible for adding the downstream domain button workflow to the Linked Projects page. This includes the following:

- Add Downstream Domain button and other minor UI tweaks to the Linked Projects page
- Passing in a list of available domains to link with when rendering the django view
- Trigger a modal to select from a list of available domains to link to. Currently only supports linking one at a time.
- Manually managing the list of both available domains and linked domains on the frontend to avoid excessive ajax calls

I left a todo in here to make an rmi call to add the new downstream domain which I will tackle next. This leaves the page in a bit of an incomplete state in that from the user's perspective it looks like they've added a domain, but they will be unable to remove it via the Remove Link button. However a simple refresh will remove them since no domains are actually being linked yet.

<img width="1191" alt="Screen Shot 2021-06-02 at 3 57 30 PM" src="https://user-images.githubusercontent.com/15785053/120545568-d208eb80-c3bc-11eb-829c-8f6cf9683498.png">
<img width="858" alt="Screen Shot 2021-06-02 at 3 57 37 PM" src="https://user-images.githubusercontent.com/15785053/120545581-d59c7280-c3bc-11eb-8e30-c8bc3eee5b04.png">
<img width="1195" alt="Screen Shot 2021-06-02 at 3 57 50 PM" src="https://user-images.githubusercontent.com/15785053/120545635-e5b45200-c3bc-11eb-969f-c775daed4789.png">

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`LINKED_DOMAINS`
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added automated tests for the method to obtain a list of available domains to link with.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Will run through QA before merging to master on parent branch.
### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Locally tested. This is not in a complete state and I have no intention of merging this PR into master as it is now.
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
initial UI work

- add downstream button
- trigger modal

gherceg committed:
add method to retrieve available domains to link

gherceg committed:
setup method to handle adding new link on the frontend

gherceg committed:
fix select2 usage

gherceg committed:
remove from availableDomains

not sure why I didn't do that before

gherceg committed:
put button behind feature flag

gherceg committed:
add href field

gherceg committed:
I'm sorry stickler

orangejenny commented on corehq/apps/linked_domain/templates/linked_domain/domain_links.html:
```diff
@@ -79,6 +79,18 @@ <h2>{% trans "Project Link" %}</h2>  {# header for the only tab that will be dis
         <div id="ko-tabs-manage-downstream" class="ko-template">
+          <h3>{% trans "Manage Downstream Projects" %}</h3>
+          <p>{% trans "This project is an upstream project space for the projects listed below." %}</p>
+          <p><a href="#">{% trans "Learn more" %}</a>{% trans " about Linked Projects." %}</p>
```
`target="_blank"` is also good for help links.

orangejenny reviewed (approved):
Just minor comments, looks good.

gherceg commented on corehq/apps/linked_domain/static/linked_domain/js/domain_links.js:
```diff
@@ -68,6 +69,7 @@ hqDefine("linked_domain/js/domain_links", [
         }
 
         self.domain_links = ko.observableArray(_.map(data.linked_domains, DomainLink));
+        self.newlyAddedDomainLinks = ko.observableArray([]);
```
This is fair. The idea is to add new domains to the top of the table, but I could achieve that with using `unshift` on the `self.domain_links` array couldn't I?

gherceg committed:
address feedback

- put text into a blocktrans
- use target="_blank" for help link
- rename test_db_accessors.py to test_dbaccessors.py
- remove newlyAddedDomainLinks since it is unnecessary

biyeun reviewed (approved):
Looks great! 🎉

gherceg committed:
update and sort availableDomains when removing and adding links

gherceg committed:
forgot a semicolon
